frappe/workflow/doctype/workflow_action/workflow_action.py

```diff
@@ -132,20 +138,95 @@ def return_link_expired_page(doc, doc_workflow_state):
 				frappe.bold(frappe.get_value('User', doc.get("modified_by"), 'full_name'))
 			), indicator_color='blue')
 
-def clear_old_workflow_actions(doc, user=None):
+
+def update_completed_workflow_actions(doc, user=None, workflow=None, workflow_state=None):
 	user = user if user else frappe.session.user
-	frappe.db.delete("Workflow Action", {
-		"reference_doctype": doc.get("doctype"),
-		"reference_name": doc.get("name"),
-		"user": ("!=", user),
-		"status": "Open"
-	})
 
-def update_completed_workflow_actions(doc, user=None):
+	allowed_roles = frappe.get_all('Workflow Transition',
+		fields='allowed',
+		filters=[['parent', '=', workflow],
+		['next_state', '=', workflow_state]],
+		pluck = 'allowed')
+
+	# There is no transaction leading upto this state
+	# so no older actions to complete
+	if not allowed_roles:
+		return
+
+	user_roles = set(frappe.get_roles(user))
+
+	allowed_roles = set(allowed_roles).intersection(user_roles)
+
+	WorkflowAction = DocType("Workflow Action")
+
+	if is_role_set(WorkflowAction, doc, allowed_roles):
+		frappe.qb.update(WorkflowAction).set(
+			WorkflowAction.status, 'Completed',
+		).set(
+			WorkflowAction.completed_by, user
+		).where(
+			WorkflowAction.reference_doctype == doc.get('doctype')
+		).where(
+			WorkflowAction.reference_name == doc.get('name')
+		).where(
+			WorkflowAction.role.isin(list(allowed_roles))
+		).where(
+			WorkflowAction.status == 'Open',
+		).where(
+			WorkflowAction.is_deleted == 0,
```

Comment:
`&` allows `AND`ing multiple criterions 

`.where((a == b) & (c != d)))`

Reply:
chaining looks cleaner to me and more pythonic
No need for brackets to separate between & too
is what you are suggesting better/

Reply:
not necessarily better but it's closer to SQL query. SQL queries only have one `where`.